e3nn/point/kernelconv.py

```diff
@@ -0,0 +1,103 @@
+import torch
+
+import e3nn.o3 as o3
+import e3nn.rs as rs
+from e3nn.kernel import Kernel
+
+
+class KernelConv(Kernel):
+    def __init__(self, Rs_in, Rs_out, RadialModel, get_l_filters=o3.selection_rule, sh=o3.spherical_harmonics_xyz, normalization='norm'):
+        """
+        :param Rs_in: list of triplet (multiplicity, representation order, parity)
+        :param Rs_out: list of triplet (multiplicity, representation order, parity)
+        :param RadialModel: Class(d), trainable model: R -> R^d
+        :param get_l_filters: function of signature (l_in, l_out) -> [l_filter]
+        :param sh: spherical harmonics function of signature ([l_filter], xyz[..., 3]) -> Y[m, ...]
+        :param normalization: either 'norm' or 'component'
+        representation order = nonnegative integer
+        parity = 0 (no parity), 1 (even), -1 (odd)
+        """
+        super(KernelConv, self).__init__(Rs_in, Rs_out, RadialModel, get_l_filters, sh, normalization)
+
+    def forward(self, features, geometry, mask, y=None, radii=None):
+        """
+        :param features: tensor [batch, b, l_in * mul_in * m_in]
+        :param geometry: tensor [batch, a, b, xyz]
+        :param mask:     tensor [batch, a] (In order to zero contributions from padded atoms.)
+        :param y:        Optional precomputed spherical harmonics.
+        :param radii:    Optional precomputed normed geometry.
+        :return:         tensor [batch, a, l_out * mul_out * m_out]
+        """
+        batch, a, b, xyz = geometry.size()
+        assert xyz == 3
+
+        # precompute all needed spherical harmonics
+        if y is None:
+            y = self.sh(self.set_of_l_filters, geometry)  # [l_filter * m_filter, batch, a, b]
+
+        # use the radial model to fix all the degrees of freedom
+        # note: for the normalization we assume that the variance of R[i] is one
+        if radii is None:
+            radii = geometry.norm(2, dim=-1)  # [batch, a, b]
+        r = self.R(radii.flatten()).view(
+            *radii.shape, -1
+        )  # [batch, a, b, l_out * l_in * mul_out * mul_in * l_filter]
+
+        norm_coef = getattr(self, 'norm_coef')
+        norm_coef = norm_coef[:, :, (radii == 0).type(torch.long)]  # [l_out, l_in, batch, a, b]
+
+        kernel_conv = kernel_conv_automatic_backward(
+            features, y, r, norm_coef, self.Rs_in, self.Rs_out, self.get_l_filters, self.set_of_l_filters
+        )
+        return kernel_conv * mask.unsqueeze(-1)
+
+
+def kernel_conv_automatic_backward(features, Y, R, norm_coef, Rs_in, Rs_out, get_l_filters, set_of_l_filters):
```

Comment:
the name is super super weird. why is is called `automatic_backward` ? I don't want to call al my function `automatic_backward` just because I don't provide the custom backward.